lib/attrsets.nix

```diff
@@ -22,12 +22,14 @@ rec {
        => 6
   */
   attrByPath = attrPath: default: e:
-    let attr = head attrPath;
-    in
-      if attrPath == [] then e
-      else if e ? ${attr}
-      then attrByPath (tail attrPath) default e.${attr}
-      else default;
+    let
+      len = length attrPath;
+      getDepth = n: attr:
+        if n == len then attr
+        else if attr ? ${elemAt attrPath n}
+        then getDepth (n + 1) attr.${elemAt attrPath n}
+        else default;
+    in getDepth 0 e;
```

Comment:
This function is strict in all of `attrPath`, so you could try `foldl'` without loss of laziness.

Reply:
The `elemAt` call is duplicated. Maybe a `let` pays off; an extra env, but two call fewer (two instead of one because of currying).

Reply:
I considered using `foldl`, but since `foldl'` is strict in both arguments, it will lose the short circuit property. Maybe I can somehow wrap the accumulator so it isn't reduced to NF, but I am not sure how to accomplish this.

Reply:
I was wrong about the strictness. `attrPath ["a" (abort "")] "def" { } => "def"`.
I only considered the non-default case.